core/hashstorage.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2019 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package core implements the Ethereum consensus protocol.
+
+package core
+
+import (
+	"bytes"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// For EVM execution, we need around 256 items. We add a few more to allow reorgs.
+// For mainnet, a couple more would suffice, but a few more added for
+// testnets/private nets
+const HashStorageElems = 288
+const HashStorageDataSize = 32 * HashStorageElems
+
+// HashStorage implements a storage for chains of hashes, intended to be used for quick lookup of blockhashes.
+// Internally, it uses an array of 8192 bytes to store 256 hashes in a circular buffer.
+// It enforces that all hashes added have a contiguous parent-child relation, and supports rollbacks
+// It is not thread-safe.
+type HashStorage struct {
+
+	// The data holds 256 32-byte hashes. The hashes are sequential, but also a
+	// circular buffer.
+	// The `head` points to the position of the latest hash.
+	// The parent, if present, is located 32 bytes back.
+	// [.., .., ..., head-2 , head-1, head, oldest, ... ]
+	data [HashStorageDataSize]byte
+
+	head uint64 // index of hash for the head block
+
+	headNumber uint64 // The blocknumber for head (the most recent block)
+	tailNumber uint64 // The blocknumber for tail (the oldest block)
+}
+
+// Create a new storage with a header in it.
+func NewHashStorage(hdr *types.Header) *HashStorage {
```

Comment:
It seems weird to me that to create a data structure, we already need an item to put into it. Why tie the two together? Can't we just have a constructor and a setter? Also, perhaps we could make the number of hashes it can contain a constructor paramter instead of hard coding it?

Reply:
+1. Also a name nit: `header` is already pretty short that `hdr` is on that shortness-readability line.

Reply:
The reason is that since it's a circular buffer, it simplifies things if it cannot be empty. Then we have to add checks to distinguish between if it's full or empty, basically.